text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+system.  The one new twist is the use of `Pin`, which makes it possible to keep
+data borrowed across separate calls to `poll` (i.e., "borrowing over yield
+points"). The mechanics of pinning are explained
+in [the RFC that introduced it](https://github.com/rust-lang/rfcs/pull/2349),
+but one key point is that when implementing `Future` by hand, if you don't wish
+to use this borrowing feature, you can treat `Pin<Self>` just like `&mut self`.
+
+### Universal adapters
+
+There are several basic adapters, modeled on `Iterator`, that work for arbitrary
+futures. Using `impl Trait` notation, we can define them as follows:
+
+```rust
+trait Future {
+    // Transform the result of the future
+    fn map<T>(self, f: impl FnOnce(Self::Item) -> T) -> impl Future<Item = T>
+        { .. }
+
+    // Chain a future onto this one
+    fn then<F>(self, f: impl FnOnce(Self::Item) -> F) -> impl Future<Item = F::Item>
+        where F: Future
+        { .. }
+
+    // Chain a closure for side effects
+    fn inspect(self, f: impl FnOnce(&Self::Item)) -> impl Future<Item = Self::Item>
+        { .. }
+
+    // Translate unwinding within this future into a `Result`
+    fn catch_unwind(self) -> impl Future<Item = Result<Self::Item, Box<Any + Send>>>
+        where Self: UnwindSafe
+        { .. }
+}
+```
+
+These adapters are all straightforward and based on substantial prior art.
+
+More interesting adapters like `shared`, `select` and `join` will be left to the
+ecosystem to iterate on before being RFC'ed for `std`.
+
+### `Result`-specific adapters
+
+Futures are often enough used with `Result` values that we provide a distinct
+subtrait for that case, equipped with some additional adapters:
+
+```rust
+trait FutureRes<T, E>: Future<Item = Result<T, E>> {
+    // Transform the successful result of the future
+    fn map_ok<U>(self, f: impl FnOnce(T) -> U) -> impl FutureRes<U, E>
+        { .. }
+
+    // Transform the error result of the future
+    fn map_err<F>(self, f: impl FnOnce(E) -> F) -> impl FutureRes<T, F>
+        { .. }
+
+    // Chain a future onto this one on success
+    fn and_then<F, U>(self, f: impl FnOnce(T) -> F) -> impl FutureRes<U, E>
+        where F: FutureRes<U, E>
+        { .. }
+
+    // Chain a future onto this one on failure
+    fn or_else<F, G>(self, f: impl FnOnce(E) -> F) -> impl FutureRes<T, G>
+        where F: FutureRes<T, G>
+        { .. }
+
+    // Pass the error type through an arbitrary conversion
+    fn err_into<F>(self) -> impl FutureRes<T, F>
+        where E: Into<F>
+        { .. }
+
+    // Handle the error provided by this future
+    fn recover<F>(self, f: impl FnOnce(E) -> F) -> impl Future<Item = T>
+        where F: Future<Item = T>
+        { .. }
+}
+
+// Automatically applied to all `Result`-returning futures
+impl<T, E, F> FutureRes<T, E> for F where F: Future<Item = Result<T, E>> {}
+```
+
+## Prelude
+
+The `Future` and `FutureRes` traits are added to the prelude.
```

Comment:
Why are they added to the prelude?

Reply:
At least, could they be moved to `std::futures::prelude` (or whatever module they will be in)?

Reply:
I have reservations about this. Could this section be extended with some rationale?

Reply:
To provide some context for prelude or not evaluation (I don't have an opinion here yet), here's a snippet from [`std::prelude`](https://doc.rust-lang.org/nightly/std/prelude/index.html):
> The prelude is the list of things that Rust automatically imports into every Rust program. It's kept as small as possible, and is focused on things, particularly traits, which are used in almost every single Rust program.

Reply:
I would say that *most* Rust programs would not need `Future` or `FutureRes`. Not even all async programs will need it (if you use mio directly, you have no need for `Future`).

The worse aspect is that I anticipate including this in `prelude` to cause problems, as I described in [this](https://github.com/rust-lang/rfcs/pull/2395#issuecomment-379430705) comment.